File: 0072-edit-distance/0072-edit-distance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:

        if len(word1) < len(word2):
            word1, word2 = word2, word1

        if len(word2) == 0:
            return len(word1)

        dp = [len(word2) - c for c in range(len(word2))]

        for r in range(len(word1) - 1, -1, -1):
            bottomRight = len(word1) - r - 1
            right = bottomRight + 1
            for c in range(len(word2) - 1, -1, -1):
                below = dp[c]
                if word1[r] == word2[c]:
                    dp[c] = bottomRight
                else:
                    replace = bottomRight
                    _delete = below
                    add = right

                    best = min(replace, _delete, add)
                    dp[c] = best + 1

                bottomRight = below
                right = dp[c]

        return dp[0]

File: 0072-edit-distance/0072-edit-distance.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        if(word1.size() < word2.size()){
            swap(word1, word2);
        }

        vector<int> dp(word2.size() + 1, 0);
        for(int c = dp.size() - 1; c >= 0; c--){
            int ways = dp.size() - 1 - c;
            dp[c] = ways;
        }

        for(int r = word1.size() - 1; r >= 0; r--){
            int bottomRight = dp.back();
            dp.back() = dp.back() + 1;
            for(int c = word2.size() - 1; c >= 0; c--){
                int bottom = dp[c];
                int right = dp[c + 1];

                int res = INT_MAX;
                if(word1[r] == word2[c]){
                    res = min(res, bottomRight);
                }   

                res = min(res, right + 1);
                res = min(res, bottom + 1);
                res = min(res, bottomRight + 1);

                bottomRight = bottom;
                dp[c] = res;
            }
        }
        return dp.front();
    }
};